File: projects/mcp-8004-enterprise/agents_implementation/credit-risk-bridge.js
/**
 * credit-risk-bridge.js  [BANK DMZ EXTERNAL — bidirectional]
 *
 * Watches CreditRiskOracle events and routes to/from the credit-risk-server.
 *
 * Event flows:
 *   CreditReviewRequested → assess_credit
 *     → request_documents → requestClientData()
 *     → propose_terms     → proposeTerms()
 *     → submit_recommendation → submitRecommendation()
 *
 *   DataFulfilled   → continue_assessment(trigger='data_fulfilled')
 *     → same actions as above
 *
 *   CounterProposed → continue_assessment(trigger='counter_proposed')
 *     → propose_terms  → proposeTerms() (new round)
 *     → accept_terms   → acceptTerms() then submitRecommendation()
 *
 * Usage:
 *   node credit-risk-bridge.js --contract 0x... --rpc ... --privkey ... --agent-id 0
 * ENV: CREDIT_CONTRACT_ADDRESS, RPC_URL, ORACLE_PRIVATE_KEY, CREDIT_AGENT_ID
 */

import { ethers } from 'ethers';
import { bootstrapBridge, callMcpTool, governancePreflight, arg } from './bridge-base.js';

const LABEL            = 'credit-risk-bridge';
const MCP_ENDPOINT     = arg('--mcp-endpoint', 'CREDIT_MCP_ENDPOINT') ?? 'http://localhost:8011';
const CAP_CREDIT       = ethers.id('credit_review');
const TOOL_ASSESS      = ethers.id('assess_credit');

const ABI = [
  'event CreditReviewRequested(bytes32 indexed requestId, bytes32 indexed flowId, uint256 bankAgentId, uint256 clientAgentId, uint256 timestamp)',
  'event DataFulfilled(bytes32 indexed requestId, bytes32 indexed flowId, bytes32 dataHash, uint256 submittingAgentId, uint256 timestamp)',
  'event CounterProposed(bytes32 indexed requestId, bytes32 indexed flowId, bytes32 proposalHash, uint256 agentId, uint256 timestamp)',
  'function requestClientData(bytes32 requestId, uint256 bankAgentId, bytes32 dataSpecHash)',
  'function proposeTerms(bytes32 requestId, uint256 bankAgentId, bytes32 termsHash)',
  'function acceptTerms(bytes32 requestId, uint256 bankAgentId, bytes32 agreedTermsHash)',
  'function submitRecommendation(bytes32 requestId, uint256 bankAgentId, bytes32 resultHash)',
  'function getRequest(bytes32 requestId) view returns (tuple(bytes32 flowId, uint256 clientAgentId, uint256 bankAgentId, uint8 status, bytes32 dataRequestSpec, uint8 dataRequestRound, bytes32 currentTermsHash, uint8 negotiationRound, bytes32 resultHash, uint256 createdAt))',
];

async function dispatchAction(oracle, agentId, requestId, result, traceId) {
  if (result.action === 'request_documents') {
    console.log(`[${LABEL}]   → requestClientData  spec=${result.spec_hash}`);
    const tx = await oracle.requestClientData(requestId, agentId, result.spec_hash);
    await tx.wait();
    console.log(`[${LABEL}]   ✓ requestClientData  tx=${tx.hash}`);
  } else if (result.action === 'propose_terms') {
    console.log(`[${LABEL}]   → proposeTerms  hash=${result.terms_hash}`);
    const tx = await oracle.proposeTerms(requestId, agentId, result.terms_hash);
    await tx.wait();
    console.log(`[${LABEL}]   ✓ proposeTerms  tx=${tx.hash}`);
  } else if (result.action === 'accept_terms') {
    console.log(`[${LABEL}]   → acceptTerms  agreed=${result.agreed_hash}`);
    const tx1 = await oracle.acceptTerms(requestId, agentId, result.agreed_hash);
    await tx1.wait();
    console.log(`[${LABEL}]   ✓ acceptTerms  tx=${tx1.hash}`);
    // After accepting, submit recommendation
    const req = await oracle.getRequest(requestId);
    const resultHash = ethers.keccak256(ethers.toUtf8Bytes(`credit-result:${req.flowId}`));
    const tx2 = await oracle.submitRecommendation(requestId, agentId, resultHash);
    await tx2.wait();
    console.log(`[${LABEL}]   ✓ submitRecommendation  tx=${tx2.hash}`);
  } else if (result.action === 'submit_recommendation') {
    console.log(`[${LABEL}]   → submitRecommendation  approved=${result.approved}`);
    const tx = await oracle.submitRecommendation(requestId, agentId, result.result_hash);
    await tx.wait();
    console.log(`[${LABEL}]   ✓ submitRecommendation  tx=${tx.hash}`);
  }
}

async function main() {
  const { wallet, contracts, agentId, CONTRACT_ADDRESS } = await bootstrapBridge(LABEL, {
    contractFlag: '--contract', contractEnv: 'CREDIT_CONTRACT_ADDRESS',
    agentIdEnv: 'CREDIT_AGENT_ID',
    cardGlob: f => f === 'bank-credit-risk-agent.json',
  });

  const oracle = new ethers.Contract(CONTRACT_ADDRESS, ABI, wallet);

  console.log(`[${LABEL}] MCP endpoint : ${MCP_ENDPOINT}`);
  console.log(`[${LABEL}] Listening for CreditRiskOracle events…\n`);

  oracle.on('CreditReviewRequested', async (...args) => {
    const event = args[args.length - 1];
    const { requestId, flowId, clientAgentId } = event.args;
    console.log(`\n[${LABEL}] ← CreditReviewRequested  requestId=${requestId}  flowId=${flowId}`);
    const ok = await governancePreflight(LABEL, { flowId, agentId, capability: CAP_CREDIT, toolHash: TOOL_ASSESS, contracts });
    if (!ok) return;
    try {
      const result = await callMcpTool(MCP_ENDPOINT, 'assess_credit', {
        flow_id: flowId, request_id: requestId, client_agent_id: clientAgentId.toString(),
      }, flowId);
      await dispatchAction(oracle, agentId, requestId, result, flowId);
    } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
  });

  oracle.on('DataFulfilled', async (...args) => {
    const event = args[args.length - 1];
    const { requestId, flowId, dataHash, submittingAgentId } = event.args;
    if (submittingAgentId === agentId) return; // our own tx
    console.log(`\n[${LABEL}] ← DataFulfilled  requestId=${requestId}`);
    const ok = await governancePreflight(LABEL, { flowId, agentId, capability: CAP_CREDIT, toolHash: TOOL_ASSESS, contracts });
    if (!ok) return;
    try {
      const req    = await oracle.getRequest(requestId);
      const result = await callMcpTool(MCP_ENDPOINT, 'continue_assessment', {
        flow_id: flowId, request_id: requestId,
        trigger: 'data_fulfilled', data_hash: dataHash,
        current_round: Number(req.negotiationRound),
      }, flowId);
      await dispatchAction(oracle, agentId, requestId, result, flowId);
    } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
  });

  oracle.on('CounterProposed', async (...args) => {
    const event = args[args.length - 1];
    const { requestId, flowId, proposalHash } = event.args;
    console.log(`\n[${LABEL}] ← CounterProposed  requestId=${requestId}  proposal=${proposalHash}`);
    const ok = await governancePreflight(LABEL, { flowId, agentId, capability: CAP_CREDIT, toolHash: TOOL_ASSESS, contracts });
    if (!ok) return;
    try {
      const req    = await oracle.getRequest(requestId);
      const result = await callMcpTool(MCP_ENDPOINT, 'continue_assessment', {
        flow_id: flowId, request_id: requestId,
        trigger: 'counter_proposed', data_hash: proposalHash,
        current_round: Number(req.negotiationRound),
      }, flowId);
      await dispatchAction(oracle, agentId, requestId, result, flowId);
    } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
  });
}

main().catch(err => { console.error(err); process.exit(1); });

File: projects/mcp-8004-enterprise/agents_implementation_py/bridges/credit_risk_bridge.py
"""
credit_risk_bridge.py — Port of credit-risk-bridge.js using web3.py AsyncWeb3.

Watches CreditRiskOracle events and routes to/from the Python credit risk MCP server.

Event flows:
  CreditReviewRequested → assess_credit
    → request_documents → requestClientData()
    → propose_terms     → proposeTerms()
    → submit_recommendation → submitRecommendation()
  DataFulfilled   → continue_assessment(trigger='data_fulfilled')
  CounterProposed → continue_assessment(trigger='counter_proposed')
    → propose_terms  → proposeTerms() (new round)
    → accept_terms   → acceptTerms() then submitRecommendation()

Usage:
    python -m bridges.credit_risk_bridge \
        --contract 0x... --rpc http://... --privkey 0x... --agent-id 1
ENV: CREDIT_CONTRACT_ADDRESS, RPC_URL, ORACLE_PRIVATE_KEY, CREDIT_AGENT_ID
"""

from __future__ import annotations

import asyncio
import logging

from web3 import Web3

from shared.abis import CREDIT_ORACLE_ABI
from shared.bridge_base import arg, bootstrap_bridge, call_mcp_tool, governance_preflight

logger = logging.getLogger(__name__)

LABEL        = "credit-risk-bridge"
MCP_ENDPOINT = arg("--mcp-endpoint", "CREDIT_MCP_ENDPOINT", "http://localhost:8011")

CAP_CREDIT  = Web3.keccak(text="credit_review")
TOOL_ASSESS = Web3.keccak(text="assess_credit")


def _hex(b) -> str:
    if isinstance(b, bytes):
        return b.hex()
    return str(b).removeprefix("0x")


def _bytes(h: str) -> bytes:
    return bytes.fromhex(h.removeprefix("0x"))


async def _dispatch(oracle, agent_id: int, request_id: bytes, result: dict) -> None:
    """Dispatch the on-chain action indicated by the MCP tool result."""
    action = result.get("action")

    if action == "request_documents":
        logger.info("[%s]   → requestClientData  spec=%s", LABEL, result["spec_hash"])
        tx = await oracle.functions.requestClientData(
            request_id, agent_id, _bytes(result["spec_hash"])
        ).transact()
        logger.info("[%s]   ✓ requestClientData  tx=%s", LABEL, tx.hex())

    elif action == "propose_terms":
        logger.info("[%s]   → proposeTerms  hash=%s", LABEL, result["terms_hash"])
        tx = await oracle.functions.proposeTerms(
            request_id, agent_id, _bytes(result["terms_hash"])
        ).transact()
        logger.info("[%s]   ✓ proposeTerms  tx=%s", LABEL, tx.hex())

    elif action == "accept_terms":
        logger.info("[%s]   → acceptTerms  agreed=%s", LABEL, result["agreed_hash"])
        tx1 = await oracle.functions.acceptTerms(
            request_id, agent_id, _bytes(result["agreed_hash"])
        ).transact()
        logger.info("[%s]   ✓ acceptTerms  tx=%s", LABEL, tx1.hex())
        # After accepting, submit recommendation
        req = await oracle.functions.getRequest(request_id).call()
        result_hash = Web3.keccak(text=f"credit-result:{_hex(req[0])}")
        tx2 = await oracle.functions.submitRecommendation(
            request_id, agent_id, result_hash
        ).transact()
        logger.info("[%s]   ✓ submitRecommendation  tx=%s", LABEL, tx2.hex())

    elif action == "submit_recommendation":
        logger.info("[%s]   → submitRecommendation  approved=%s", LABEL, result.get("approved"))
        tx = await oracle.functions.submitRecommendation(
            request_id, agent_id, _bytes(result["result_hash"])
        ).transact()
        logger.info("[%s]   ✓ submitRecommendation  tx=%s", LABEL, tx.hex())


async def _handle_review_requested(oracle, agent_id: int, contracts: dict, event) -> None:
    args = event["args"]
    request_id, flow_id = args["requestId"], args["flowId"]
    client_agent_id     = args["clientAgentId"]
    trace_id = _hex(flow_id)

    logger.info("\n[%s] ← CreditReviewRequested  requestId=%s  flowId=%s", LABEL, _hex(request_id), trace_id)

    if not await governance_preflight(
        LABEL, flow_id=flow_id, agent_id=agent_id,
        capability=CAP_CREDIT, tool_hash=TOOL_ASSESS, contracts=contracts,
    ):
        return

    try:
        result = await call_mcp_tool(
            MCP_ENDPOINT, "assess_credit",
            {"flow_id": trace_id, "request_id": _hex(request_id), "client_agent_id": str(client_agent_id)},
            trace_id,
        )
        await _dispatch(oracle, agent_id, request_id if isinstance(request_id, bytes) else _bytes(_hex(request_id)), result)
    except Exception as exc:
        logger.error("[%s]   ✗ %s", LABEL, exc)


async def _handle_data_fulfilled(oracle, agent_id: int, contracts: dict, event) -> None:
    args = event["args"]
    request_id, flow_id = args["requestId"], args["flowId"]
    data_hash        = args["dataHash"]
    submitting_agent = args["submittingAgentId"]
    trace_id         = _hex(flow_id)

    if submitting_agent == agent_id:
        return

    logger.info("\n[%s] ← DataFulfilled  requestId=%s", LABEL, _hex(request_id))

    if not await governance_preflight(
        LABEL, flow_id=flow_id, agent_id=agent_id,
        capability=CAP_CREDIT, tool_hash=TOOL_ASSESS, contracts=contracts,
    ):
        return

    try:
        req    = await oracle.functions.getRequest(request_id).call()
        result = await call_mcp_tool(
            MCP_ENDPOINT, "continue_assessment",
            {
                "flow_id": trace_id, "request_id": _hex(request_id),
                "trigger": "data_fulfilled", "data_hash": _hex(data_hash),
                "current_round": int(req[7]),  # negotiationRound
            },
            trace_id,
        )
        req_bytes = request_id if isinstance(request_id, bytes) else _bytes(_hex(request_id))
        await _dispatch(oracle, agent_id, req_bytes, result)
    except Exception as exc:
        logger.error("[%s]   ✗ %s", LABEL, exc)


async def _handle_counter_proposed(oracle, agent_id: int, contracts: dict, event) -> None:
    args = event["args"]
    request_id, flow_id = args["requestId"], args["flowId"]
    proposal_hash = args["proposalHash"]
    trace_id      = _hex(flow_id)

    logger.info("\n[%s] ← CounterProposed  requestId=%s  proposal=%s", LABEL, _hex(request_id), _hex(proposal_hash))

    if not await governance_preflight(
        LABEL, flow_id=flow_id, agent_id=agent_id,
        capability=CAP_CREDIT, tool_hash=TOOL_ASSESS, contracts=contracts,
    ):
        return

    try:
        req    = await oracle.functions.getRequest(request_id).call()
        result = await call_mcp_tool(
            MCP_ENDPOINT, "continue_assessment",
            {
                "flow_id": trace_id, "request_id": _hex(request_id),
                "trigger": "counter_proposed", "data_hash": _hex(proposal_hash),
                "current_round": int(req[7]),
            },
            trace_id,
        )
        req_bytes = request_id if isinstance(request_id, bytes) else _bytes(_hex(request_id))
        await _dispatch(oracle, agent_id, req_bytes, result)
    except Exception as exc:
        logger.error("[%s]   ✗ %s", LABEL, exc)


async def main() -> None:
    ctx = await bootstrap_bridge(
        LABEL,
        contract_flag="--contract",
        contract_env="CREDIT_CONTRACT_ADDRESS",
        agent_id_env="CREDIT_AGENT_ID",
        card_glob="bank-credit-risk-agent.json",
    )

    oracle = ctx.w3.eth.contract(address=ctx.contract_address, abi=CREDIT_ORACLE_ABI)
    logger.info("[%s] MCP endpoint : %s", LABEL, MCP_ENDPOINT)
    logger.info("[%s] Listening for CreditRiskOracle events…", LABEL)

    review_filter   = await oracle.events.CreditReviewRequested.create_filter(from_block="latest")
    data_filter     = await oracle.events.DataFulfilled.create_filter(from_block="latest")
    counter_filter  = await oracle.events.CounterProposed.create_filter(from_block="latest")

    while True:
        for event in await review_filter.get_new_entries():
            asyncio.create_task(_handle_review_requested(oracle, ctx.agent_id, ctx.contracts, event))
        for event in await data_filter.get_new_entries():
            asyncio.create_task(_handle_data_fulfilled(oracle, ctx.agent_id, ctx.contracts, event))
        for event in await counter_filter.get_new_entries():
            asyncio.create_task(_handle_counter_proposed(oracle, ctx.agent_id, ctx.contracts, event))
        await asyncio.sleep(2)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(main())
